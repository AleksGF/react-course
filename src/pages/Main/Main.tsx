import React, { type FC, useEffect, useRef, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import { fetchPeople } from '@services/api/fetchPeople';
import { getExtendedSearchParams } from '@helpers/getExtendedSearchParams';
import NavBar from '@components/NavBar/NavBar';
import PersonDetails from '@components/PersonDetails/PersonDetails';
import type { MainProps } from '@types/types';
import type { Person } from '@types/apiTypes';

const Main: FC<MainProps> = (props) => {
  const {
    searchValue,
    shouldUpdateData,
    setIsLoading,
    setSearchValue,
    setShouldUpdateData,
  } = props;
  const [searchParams, setSearchParams] = useSearchParams();
  const [peopleToShow, setPeopleToShow] = useState<Person[]>([]);
  const totalPeopleCount = useRef(0);

  const isDetailsActive = Boolean(searchParams.get('details'));

  useEffect(() => {
    const pageFromSearchParams = searchParams.get('page');

    if (!pageFromSearchParams) {
      setSearchParams(getExtendedSearchParams(searchParams, { page: '1' }));
    }
  }, [searchParams, setSearchParams]);

  useEffect(() => {
    if (shouldUpdateData && (searchValue || searchValue === '')) {
      const search = searchValue.trim();
      localStorage.setItem('rc_lastSearch', search);
      const currentPage = Number(searchParams.get('page') ?? '1');
      const personsPerPage = Number(searchParams.get('limit') ?? '10');

      setIsLoading(true);

      fetchPeople(currentPage, personsPerPage === 20 ? personsPerPage : 10, {
        search,
      }).then((data) => {
        setPeopleToShow(data.people);
        totalPeopleCount.current = data.totalCount;
        setSearchValue(search);
        setIsLoading(false);
      });

      setShouldUpdateData(false);
    }
  }, [
    shouldUpdateData,
    searchValue,
    searchParams,
    setIsLoading,
    setShouldUpdateData,
    setSearchValue,
  ]);

  return (
    <main className={'content-wrapper'}>
      <NavBar
        people={peopleToShow}
        totalPeopleCount={totalPeopleCount}
        setShouldUpdateData={setShouldUpdateData}
      />
      {isDetailsActive && <PersonDetails />}
    </main>
  );
};

export default Main;
