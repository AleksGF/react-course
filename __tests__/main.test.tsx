/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom/';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import mockRouter from 'next-router-mock';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';
import { server } from '@src/tests/__mocks__/mockServer';
import Main from '@pages/main/index';
import {
  apiListPageOne,
  apiPerson,
  personsFromFirstPage,
  searchPersonName,
} from '@src/tests/__mocks__/mockApiData';
import {
  FIRST_PAGE,
  ITEMS_PER_PAGE,
  RoutePath,
  SearchParamsKeys,
} from '@src/constants/constants';

const user = userEvent.setup();

jest.mock('next/router', () => jest.requireActual('next-router-mock'));

const mockUseRouter = jest.spyOn(require('next/router'), 'useRouter');

const mockRouterPush = jest.fn();

const mockRouterImplementation = {
  query: { [SearchParamsKeys.SEARCH_VALUE]: '' },
  asPath: `${RoutePath.MAIN}?${SearchParamsKeys.PAGE_NUMBER}=${FIRST_PAGE}&${SearchParamsKeys.ITEMS_PER_PAGE}=${ITEMS_PER_PAGE.DEFAULT}`,
  push: mockRouterPush,
};

describe('Main', () => {
  beforeAll(() => server.listen());
  beforeEach(() => mockRouterPush.mockClear());
  afterAll(() => server.close());

  it('renders correctly', () => {
    const view = render(<Main people={apiListPageOne} />);

    expect(view).toMatchSnapshot();
  });

  it('renders correctly with details', () => {
    const view = render(<Main people={apiListPageOne} details={apiPerson} />);

    expect(view).toMatchSnapshot();
  });

  it('renders main search input, button, select, pagination, details', () => {
    render(<Main people={apiListPageOne} details={apiPerson} />);

    expect(screen.getByTestId('searchBar')).toBeInTheDocument();
    expect(screen.getByTestId('searchInput')).toBeInTheDocument();
    expect(screen.getByTestId('searchButton')).toBeInTheDocument();
    expect(screen.getByTestId('select')).toBeInTheDocument();
    expect(screen.getByTestId('itemsList')).toBeInTheDocument();
    expect(screen.getByTestId('pagination__btn_prev')).toBeInTheDocument();
    expect(screen.getByTestId('pagination__page')).toBeInTheDocument();
    expect(screen.getByTestId('pagination__btn_next')).toBeInTheDocument();
    expect(screen.getByTestId('details')).toBeInTheDocument();
    expect(screen.getByTestId('details-close-btn')).toBeInTheDocument();
  });

  it('handles input text', async () => {
    render(<Main people={apiListPageOne} />);

    const inputField = screen.getByTestId('searchInput') as HTMLInputElement;

    await user.type(inputField, searchPersonName);

    expect(inputField.value).toBe(searchPersonName);
  });

  it('handles search button', async () => {
    const mockLocalStorageSetItem = jest
      .spyOn(Storage.prototype, 'setItem')
      .mockImplementation(() => jest.fn());

    mockUseRouter.mockImplementation(() => mockRouterImplementation);

    render(<Main people={apiListPageOne} />);

    const inputField = screen.getByTestId('searchInput');
    const searchButton = screen.getByTestId('searchButton');

    expect(mockLocalStorageSetItem.mock.calls).toHaveLength(0);
    expect(mockRouterPush.mock.calls).toHaveLength(0);

    await user.type(inputField, searchPersonName);
    await user.click(searchButton);

    expect(mockLocalStorageSetItem.mock.calls).toHaveLength(1);
    expect(mockRouterPush.mock.calls).toHaveLength(1);

    expect(mockRouterPush.mock.calls[0][0]).toMatch(
      `${SearchParamsKeys.SEARCH_VALUE}=${searchPersonName}`,
    );
    expect(mockRouterPush.mock.calls[0][0]).toMatch(
      `${SearchParamsKeys.ITEMS_PER_PAGE}=${ITEMS_PER_PAGE.DEFAULT}`,
    );
    expect(mockRouterPush.mock.calls[0][0]).toMatch(
      `${SearchParamsKeys.PAGE_NUMBER}=${FIRST_PAGE}`,
    );
  });

  it('handles select options', async () => {
    mockUseRouter.mockImplementation(() => mockRouterImplementation);

    render(<Main people={apiListPageOne} />);

    const selectEl = screen.getByTestId('select');

    expect(mockRouterPush.mock.calls).toHaveLength(0);

    await user.selectOptions(selectEl, String(ITEMS_PER_PAGE.DOUBLE));

    expect(mockRouterPush.mock.calls).toHaveLength(1);

    await user.selectOptions(selectEl, String(ITEMS_PER_PAGE.DEFAULT));

    expect(mockRouterPush.mock.calls).toHaveLength(2);

    expect(mockRouterPush.mock.calls[0][0]).toMatch(
      `${SearchParamsKeys.ITEMS_PER_PAGE}=${ITEMS_PER_PAGE.DOUBLE}`,
    );
    expect(mockRouterPush.mock.calls[1][0]).toMatch(
      `${SearchParamsKeys.ITEMS_PER_PAGE}=${ITEMS_PER_PAGE.DEFAULT}`,
    );
  });

  it('handles pagination', async () => {
    mockUseRouter.mockImplementation(() => mockRouterImplementation);

    render(<Main people={apiListPageOne} />);

    const prevButton = screen.getByTestId('pagination__btn_prev');
    const nextButton = screen.getByTestId('pagination__btn_next');

    expect(screen.getByText('Page# 1 from 9')).toBeInTheDocument();
    expect(prevButton).toHaveAttribute('disabled');
    expect(nextButton).not.toHaveAttribute('disabled');
    expect(mockRouterPush.mock.calls).toHaveLength(0);

    await user.click(nextButton);

    expect(mockRouterPush.mock.calls).toHaveLength(1);
    expect(mockRouterPush.mock.calls[0][0]).toMatch(
      `${SearchParamsKeys.PAGE_NUMBER}=${FIRST_PAGE + 1}`,
    );
  });

  it('handles person list click', async () => {
    render(<Main people={apiListPageOne} />, { wrapper: MemoryRouterProvider });

    for (let i = 0; i < 10; i += 1) {
      const item = screen.getByText(personsFromFirstPage[i]);

      expect(item).toBeInTheDocument();

      await user.click(item);

      expect(mockRouter.asPath).toMatch(
        `${SearchParamsKeys.DETAILS_VIEW_ID}=${i + 1}`,
      );
    }
  });

  it('handle person details close', async () => {
    mockUseRouter.mockImplementation(() => mockRouterImplementation);
    render(<Main people={apiListPageOne} details={apiPerson} />);

    const closeBtn = screen.getByTestId('details-close-btn');

    expect(closeBtn).toBeInTheDocument();
    expect(mockRouterPush.mock.calls).toHaveLength(0);

    await user.click(closeBtn);

    expect(mockRouterPush.mock.calls).toHaveLength(1);
  });
});
