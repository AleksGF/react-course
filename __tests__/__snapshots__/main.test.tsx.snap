// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header>
        <div
          class="searchBar__wrapper"
          data-testid="searchBar"
        >
          <form
            class="searchBar__form"
            name="search-field-form"
          >
            <input
              data-testid="searchInput"
              name="inputField"
              type="text"
              value=""
            />
            <button
              data-testid="searchButton"
            >
              Search
            </button>
          </form>
        </div>
      </header>
      <main>
        <h1
          class="nav__title"
        >
          Star War Persons
        </h1>
        <div
          class="select__wrapper"
        >
          <div>
            Persons per page:
          </div>
          <select
            data-testid="select"
            name="persons-per-page"
          >
            <option
              selected=""
              value="10"
            >
              10
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
        <div
          data-testid="itemsList"
        >
          <a
            class="nav-item__wrapper"
            href="/loading?details=1"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Luke Skywalker
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=2"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              C-3PO
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=3"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              R2-D2
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=4"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Darth Vader
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=5"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Leia Organa
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=6"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Owen Lars
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=7"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Beru Whitesun lars
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=8"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              R5-D4
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=9"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Biggs Darklighter
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=10"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Obi-Wan Kenobi
            </div>
          </a>
          <div
            data-testid="pagination"
          >
            <button
              data-testid="pagination__btn_prev"
              disabled=""
            >
              &lt;&lt;&lt;
            </button>
            <div
              data-testid="pagination__page"
            >
              Page# 1 from 9
            </div>
            <button
              data-testid="pagination__btn_next"
            >
              &gt;&gt;&gt;
            </button>
          </div>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <header>
      <div
        class="searchBar__wrapper"
        data-testid="searchBar"
      >
        <form
          class="searchBar__form"
          name="search-field-form"
        >
          <input
            data-testid="searchInput"
            name="inputField"
            type="text"
            value=""
          />
          <button
            data-testid="searchButton"
          >
            Search
          </button>
        </form>
      </div>
    </header>
    <main>
      <h1
        class="nav__title"
      >
        Star War Persons
      </h1>
      <div
        class="select__wrapper"
      >
        <div>
          Persons per page:
        </div>
        <select
          data-testid="select"
          name="persons-per-page"
        >
          <option
            selected=""
            value="10"
          >
            10
          </option>
          <option
            value="20"
          >
            20
          </option>
        </select>
      </div>
      <div
        data-testid="itemsList"
      >
        <a
          class="nav-item__wrapper"
          href="/loading?details=1"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Luke Skywalker
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=2"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            C-3PO
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=3"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            R2-D2
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=4"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Darth Vader
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=5"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Leia Organa
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=6"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Owen Lars
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=7"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Beru Whitesun lars
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=8"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            R5-D4
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=9"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Biggs Darklighter
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=10"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Obi-Wan Kenobi
          </div>
        </a>
        <div
          data-testid="pagination"
        >
          <button
            data-testid="pagination__btn_prev"
            disabled=""
          >
            &lt;&lt;&lt;
          </button>
          <div
            data-testid="pagination__page"
          >
            Page# 1 from 9
          </div>
          <button
            data-testid="pagination__btn_next"
          >
            &gt;&gt;&gt;
          </button>
        </div>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Main renders correctly with details 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header>
        <div
          class="searchBar__wrapper"
          data-testid="searchBar"
        >
          <form
            class="searchBar__form"
            name="search-field-form"
          >
            <input
              data-testid="searchInput"
              name="inputField"
              type="text"
              value=""
            />
            <button
              data-testid="searchButton"
            >
              Search
            </button>
          </form>
        </div>
      </header>
      <main>
        <h1
          class="nav__title"
        >
          Star War Persons
        </h1>
        <div
          class="select__wrapper"
        >
          <div>
            Persons per page:
          </div>
          <select
            data-testid="select"
            name="persons-per-page"
          >
            <option
              selected=""
              value="10"
            >
              10
            </option>
            <option
              value="20"
            >
              20
            </option>
          </select>
        </div>
        <div
          data-testid="itemsList"
        >
          <a
            class="nav-item__wrapper"
            href="/loading?details=1"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Luke Skywalker
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=2"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              C-3PO
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=3"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              R2-D2
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=4"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Darth Vader
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=5"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Leia Organa
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=6"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Owen Lars
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=7"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Beru Whitesun lars
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=8"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              R5-D4
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=9"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Biggs Darklighter
            </div>
          </a>
          <a
            class="nav-item__wrapper"
            href="/loading?details=10"
          >
            <div
              class="nav-item__name nav-link"
              data-testid="person-item"
            >
              Obi-Wan Kenobi
            </div>
          </a>
          <div
            data-testid="pagination"
          >
            <button
              data-testid="pagination__btn_prev"
              disabled=""
            >
              &lt;&lt;&lt;
            </button>
            <div
              data-testid="pagination__page"
            >
              Page# 1 from 9
            </div>
            <button
              data-testid="pagination__btn_next"
            >
              &gt;&gt;&gt;
            </button>
          </div>
        </div>
      </main>
      <div
        class="details__wrapper"
        data-testid="details"
        id="details"
      >
        <img
          alt="Close details"
          class="details__btn_close"
          data-nimg="1"
          data-testid="details-close-btn"
          decoding="async"
          height="24"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
          style="color: transparent;"
          width="24"
        />
        <div
          class="details__content"
        >
          <h3>
            Person Details:
          </h3>
          <p>
            Luke Skywalker was born in 19BBY.
          </p>
          <p>
            Films count with is 4.
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <header>
      <div
        class="searchBar__wrapper"
        data-testid="searchBar"
      >
        <form
          class="searchBar__form"
          name="search-field-form"
        >
          <input
            data-testid="searchInput"
            name="inputField"
            type="text"
            value=""
          />
          <button
            data-testid="searchButton"
          >
            Search
          </button>
        </form>
      </div>
    </header>
    <main>
      <h1
        class="nav__title"
      >
        Star War Persons
      </h1>
      <div
        class="select__wrapper"
      >
        <div>
          Persons per page:
        </div>
        <select
          data-testid="select"
          name="persons-per-page"
        >
          <option
            selected=""
            value="10"
          >
            10
          </option>
          <option
            value="20"
          >
            20
          </option>
        </select>
      </div>
      <div
        data-testid="itemsList"
      >
        <a
          class="nav-item__wrapper"
          href="/loading?details=1"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Luke Skywalker
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=2"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            C-3PO
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=3"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            R2-D2
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=4"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Darth Vader
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=5"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Leia Organa
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=6"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Owen Lars
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=7"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Beru Whitesun lars
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=8"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            R5-D4
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=9"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Biggs Darklighter
          </div>
        </a>
        <a
          class="nav-item__wrapper"
          href="/loading?details=10"
        >
          <div
            class="nav-item__name nav-link"
            data-testid="person-item"
          >
            Obi-Wan Kenobi
          </div>
        </a>
        <div
          data-testid="pagination"
        >
          <button
            data-testid="pagination__btn_prev"
            disabled=""
          >
            &lt;&lt;&lt;
          </button>
          <div
            data-testid="pagination__page"
          >
            Page# 1 from 9
          </div>
          <button
            data-testid="pagination__btn_next"
          >
            &gt;&gt;&gt;
          </button>
        </div>
      </div>
    </main>
    <div
      class="details__wrapper"
      data-testid="details"
      id="details"
    >
      <img
        alt="Close details"
        class="details__btn_close"
        data-nimg="1"
        data-testid="details-close-btn"
        decoding="async"
        height="24"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
        style="color: transparent;"
        width="24"
      />
      <div
        class="details__content"
      >
        <h3>
          Person Details:
        </h3>
        <p>
          Luke Skywalker was born in 19BBY.
        </p>
        <p>
          Films count with is 4.
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
